feat: add visibility api to control unpop

Pacman99 opened the pull request:
use to generalize __meta__ and make it invisible

The default visibility of every attribute is visible. You can set that attributes name to `false` to cause it to be removed when unpopped. A later pop can set it to `true` to make it visible again.

fare commented on POP.nix:
```diff
@@ -97,7 +97,7 @@
     defaults = lib.foldr mergeInstance bottomInstance ([meta.defaults] ++ map getDefaults precedenceList);
     __meta__ = meta // {inherit precedenceList;};
-    proto = composeProtos ([(topProto __meta__) (extensionProto meta.extension)] ++ (map getProto precedenceList));
+    proto = composeProtos ([(extensionProto meta.extension)] ++ (map getProto precedenceList) ++ [(topProto __meta__)]);
```
If moved at the end, isn't it a `bottomProto`, then?

Pacman99 replied:
yeah makes sense, I renamed it

fare commented on POP.nix:
```diff
@@ -222,7 +222,19 @@
+      super
+      // {
+        __meta__ =
+          __meta__
```
Are you sure that the `__meta__` here isn't circularly defined, vs referring to the previous binding?

Pacman99 replied:
The previous line was just `inherit __meta__`, so this should have the same effect, but with some additions

fare commented on POP.nix:
```diff
@@ -351,7 +365,15 @@
   # unpop :: Pop A B -> A
-  unpop = p: builtins.removeAttrs p ["__meta__"];
+  unpop = p:
+    builtins.removeAttrs p (
```
The `removeAttrs` becomes redundant if you already have `"__meta__"` in your bottom visibility above.

fare replied:
... assuming you either include the `bottomProto` in the list, or use its visibility as the bottom of your `foldr`.

fare replied:
If we have the function `unpop`, do we still want the field `__unpop__` by default?

Pacman99 replied:
Yeah the bottomProto is in the list, so __meta__ will be removed during an unpop. I still think `__unpop__` is useful, because it lets pops customize their unpop function. And for code that exports a pop, a consumer doesn't have to import the pop library to unpop it, the unpopping logic is embedded in it.

Pacman99 commented:
Updated to just add visibility api and not `__unpop__`. Also switched back to `topProto` and put that actually at the top.

So this PR is a lot simpler, it just adds the visibility api and changes the `unpop` function to respect it.

Pacman99 commented:
Don't really have need for functionality anymore
